modules/util/rotatingfilewriter/writer.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package rotatingfilewriter
+
+import (
+	"bufio"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/graceful/releasereopen"
+	"code.gitea.io/gitea/modules/util"
+)
+
+type Options struct {
+	Rotate           bool
+	MaximumSize      int64
+	RotateDaily      bool
+	KeepDays         int
+	Compress         bool
+	CompressionLevel int
+}
+
+type RotatingFileWriter struct {
+	mu sync.Mutex
+	fd *os.File
+
+	currentSize int64
+	openDate    int
+
+	options Options
+
+	cancelReleaseReopen func()
+}
+
+// Open creates a new rotating file writer.
+// Notice: if a file is opened by two rotators, there will be conflicts when rotating.
+// In the future, there should be "rotating file manager"
+func Open(filename string, options *Options) (*RotatingFileWriter, error) {
+	if options == nil {
+		options = &Options{}
+	}
+
+	rfw := &RotatingFileWriter{
+		options: *options,
+	}
+
+	if err := rfw.open(filename); err != nil {
+		return nil, err
+	}
+
+	rfw.cancelReleaseReopen = releasereopen.GetManager().Register(rfw)
+	return rfw, nil
+}
+
+func (rfw *RotatingFileWriter) Write(b []byte) (int, error) {
+	if rfw.options.Rotate && ((rfw.options.MaximumSize > 0 && rfw.currentSize >= rfw.options.MaximumSize) || (rfw.options.RotateDaily && time.Now().Day() != rfw.openDate)) {
+		if err := rfw.DoRotate(); err != nil {
+			// This should be
+			// return 0, err
+			// but the old behaviour does not return. This may lead to other errors.
+			fmt.Fprintf(os.Stderr, "RotatingFileWriter: %s\n", err)
```

Comment:
As this PR is breaking, could you change this too? I did not change it in the audit PR because that one should not be breaking.

Reply:
Sure, I can try to improve it. Just a moment please.

The log writer will use fallback (stderr) to output logs if the file writer returns error, then no log would be lost.

---

Actually, the log rotator code has other problems.

The code was as old as 2014 ..... maybe at that time it works fine. But now if we use two rotating file writers to write to a single file (eg: gitea.log), they just conflict, because each writer has its internal bytes counter.

IMO this problem should be resolved by something like "shared/managed rotating file writer", while we can leave the problem to future.